kernel/src/hil/hmac.rs

```diff
@@ -0,0 +1,39 @@
+//! Interface for HMAC
+
+use crate::returncode::ReturnCode;
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a> {
+    /// On error `result` will contain device specific error codes
+    fn hash_done(&'a self, result: Result<(), u32>);
+}
+
+pub trait HMAC<'a> {
+    /// Set the client instance which will receive `hash_done()` callbacks
+    fn set_client(&'a self, client: &'a dyn Client<'a>);
+
+    /// Set the hash key
+    fn set_key(&self, key: *const u8, len: usize) -> ReturnCode;
+
+    /// Add data to the hmac IP.
+    /// All data passed in is fed to the HMAC hardware block.
+    /// Returns the number of bytes written on success
+    fn add_data(&self, data: *const u8, length: usize) -> Result<usize, ReturnCode>;
+
+    /// Request the hardware block to generate a HMAC
+    /// This doesn't return anything, instead the client needs to have
+    /// set a `hash_done` handler.
+    fn run(&'a self);
+
+    /// Stores the returned digest in the memory location specified
+    fn get_digest(&self, digest: *mut u8, len: usize) -> ReturnCode;
```

Comment:
Why can't we just provide the hash to the client in the `hash_done()` callback?

Reply:
I thought about that, but I was worried the app would not just want the data in the callback. This was it's more flexible.